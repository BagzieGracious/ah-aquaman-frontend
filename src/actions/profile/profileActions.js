import { toast } from "react-toastify";
import {
  UPDATE_SUCCESS, UPDATE_FAILED, PROFILE_LOADING, IS_UPDATING, PROFILE_LOADED, PROFILE_LOAD_FAILED,
} from "../types";

const BASE_URL = process.env.BASE_URL;
export const loadProfile = () => (dispatch) => {
  dispatch({
    type: PROFILE_LOADING,
  });
  return fetch(`${BASE_URL}/profiles/${localStorage.getItem("username")}`)
    .then(res => res.json())
    .then((res) => {
      if (res.profile) {
        dispatch({
          type: PROFILE_LOADED,
          payload: res.profile,
        });
      } else {
        dispatch({
          type: PROFILE_LOAD_FAILED,
          payload: res.errors,
        });
      }
    });
};

export const updateProfile = (bio, type, history) => (dispatch) => {
  dispatch({
    type: IS_UPDATING,
  });

  const body = type === "bio" ? JSON.stringify(bio) : JSON.stringify({ image: bio });
  return fetch(`${BASE_URL}/profiles/${localStorage.getItem("username")}/edit`, {
    method: "PATCH",
    headers: {
      Authorization: `Bearer ${localStorage.getItem("token")}`,
      "Content-type": "application/json",
    },
    body,
  })
    .then(response => response.json())
    .then((data) => {
      if (data.profile) {
        (history) ? history.push("/profile") : null;
        toast.success(`${type} update success`, {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 2000,
          hideProgressBar: true,
        });
        dispatch({
          type: UPDATE_SUCCESS,
          payload: data.profile,
        });
      } else {
        dispatch({
          type: UPDATE_FAILED,
          payload: data.errors,
        });
      }
    });
};
