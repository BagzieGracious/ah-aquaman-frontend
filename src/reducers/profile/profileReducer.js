import {
  PROFILE_LOADING, GET_PROFILE,
  PROFILE_LOADED, PROFILE_LOAD_FAILED, IS_UPDATING, UPDATE_SUCCESS, UPDATE_FAILED,
} from "../../actions/types";

const initialState = {
  profile: null,
  isLoading: null,
  msg: null,
  isUpdating: null,
};
export default function (state = initialState, action) {
  switch (action.type) {
  case PROFILE_LOADED:
    return {
      ...state,
      profile: action.payload,
      isLoading: null,
      isUpdating: null,
    };
  case UPDATE_SUCCESS:
    return {
      ...state,
      isLoading: null,
      isUpdating: null,
      profile: action.payload,
    };
  case UPDATE_FAILED:
    return {
      ...state,
      msg: action.payload,
      isLoading: null,
      isUpdating: null,
    };
  case GET_PROFILE:
    return {
      ...state,
      profile: action.payload,
    };
  case PROFILE_LOADING:
    return {
      ...state,
      isLoading: true,
      isUpdating: null,
    };
  case PROFILE_LOAD_FAILED:
    return {
      ...state,
      msg: action.payload,
    };
  case IS_UPDATING:
    return {
      ...state,
      isUpdating: true,
    };
  default:
    return state;
  }
}
