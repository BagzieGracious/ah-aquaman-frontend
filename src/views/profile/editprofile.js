import React from "react";
import Progress from "./Progress";
import UpdateForm from "./UpdateForm";

const EditProfileView = (props) => {
  const {
    isUpdating, onSubmit, onChange, profile, state,
    handleChange,
  } = props;
  return (
    <div className="container">
      <div className="card mt-4">
        <div className="card-header">
          Edit Profile
        </div>
        <div className="card-body">
          <div className="row">
            <div className="col-md-2 d-none d-sm-block">
              <ul className="list-group">
                <li className="list-group-item bg-light"><strong>Edit Profile</strong></li>
                <li className="list-group-item">Change Password</li>
              </ul>
            </div>
            <div className="col-md-10">
              <div className="container">
                <div className="row">
                  <div className="col-md-2" />
                  <div className="col-md-10">
                    <div className="row">
                      <div className="col-md-2">
                        <img src={state.url ? state.url : "https://bit.ly/2Ghxc2g"} height={50} width={50} alt="dp" className="rounded" />
                      </div>
                      <div className="col-md-10">
                        <span className="lead">
                          {profile ? (
                            <strong>
                              {profile.email}
                            </strong>
                          ) : ""}
                        </span>
                        <br />
                        <input type="file" name="FileInput" className="btn btn-block siteColor bg-light" accept="image/*" onChange={handleChange} />

                        <p className="text-primary">Change Profile picture</p>
                        {state.isUploading ? (
                          <Progress percentage={state.progress} />
                        ) : ""}
                      </div>
                    </div>
                    <UpdateForm
                      isUpdating={isUpdating}
                      profile={profile}
                      onSubmit={onSubmit}
                      onChange={onChange}
                    />
                  </div>
                  <div className="col-md-3" />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default EditProfileView;
