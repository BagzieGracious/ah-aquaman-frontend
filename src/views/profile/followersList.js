import React from "react";
import Moment from "react-moment";
import ReactTimeAgo from "react-time-ago";

const FollowersList = (props) => {
  const { followersList } = props;
  return (
    <div className="modal-dialog" role="document">
      <div className="modal-content">
        <div className="modal-header">
          <h5 className="modal-title" id="exampleModalLabel">The Followers List</h5>
          <button type="button" className="close" data-dismiss="modal" aria-label="Close">
            <span aria-hidden="true">&times;</span>
          </button>
        </div>
        <div className="modal-body list-group">
          {followersList && followersList.length > 0 ? (
            followersList.map(item => (
              <div
                onClick={() => {
                  document.location.href = `/profile/${item.username}`;
                }}
                key={item.username}
                className="flex-column align-items-start"
              >
                <div className="d-flex w-100 justify-content-between">
                  <div className="row">
                    <div className="col-md-4">
                      {" "}
                      <img src={item.image ? `${item.image}` : "https://image.flaticon.com/icons/png/512/17/17004.png"} alt="" className="ourdp2 rounded img-fluid " />
                    </div>
                    <div className="col-md-8">
                      {" "}
                      <h5 className="mb-1">{item.username}</h5>
                      <p className="mb-1 text-truncate">{item.bio}</p>
                    </div>
                  </div>
                  <small>{<Moment fromNow>{item.followed_at}</Moment>}</small>
                </div>

              </div>


            ))

          ) : "No one is on the followers List"
          }


        </div>
      </div>
    </div>

  );
};

export default FollowersList;
