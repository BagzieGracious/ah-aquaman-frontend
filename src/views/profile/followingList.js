import React from "react";
import Moment from "react-moment";
import ReactTimeAgo from "react-time-ago";

const FollowingList = (props) => {
  const { list } = props;
  return (
    <div className="modal-dialog" role="document">
      <div className="modal-content">
        <div className="modal-header">
          <h5 className="modal-title" id="exampleModalLabel">The Following List</h5>
          <button type="button" className="close" data-dismiss="modal" aria-label="Close">
            <span aria-hidden="true">&times;</span>
          </button>
        </div>
        <div className="modal-body list-group">
          {list && list.length > 0 ? (
            list.map(item => (
              <div
                onClick={() => {
                  document.location.href = `/profile/${item.username}`;
                }}
                key={item.username}
                className="flex-column align-items-start"
              >
                <div className="d-flex w-100 justify-content-between">
                  <div className="row">
                    <div className="col-md-4">
                      <img src={item.image ? `${item.image}` : "https://image.flaticon.com/icons/png/512/17/17004.png"} alt="" className="ourdp2 rounded img-fluid " />
                    </div>
                    <div className="col-md-8 float-left">
                      <h5 className="mb-1 text-truncate">{item.username}</h5>
                      <p className="mb-1 text-truncate">{item.bio ? item.bio : "No bio set"}</p>
                    </div>
                  </div>
                  <small>{<Moment fromNow>{item.followed_at}</Moment>}</small>
                </div>
              </div>
            ))
          ) : "No one is on the following List"
          }
        </div>
      </div>
    </div>
  );
};

export default FollowingList;
