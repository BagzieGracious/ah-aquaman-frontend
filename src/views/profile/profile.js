import React from "react";
import { Link } from "react-router-dom";
import ProfileStats from "./ProfileStats";
import Spinner from "./Spinner";
import Sidebar from "../../components/article/sidebar";
import Navbar from "../../components/article/navbarComponent/Navbar";
import ProfileButton from "./profileButton";

const ProfileView = (props) => {
  const {
 status, profile, unFollowUser, following, followers, isFollowing, bookmarks
  } = props;

  return (
    <React.Fragment>
      <Navbar />
      <div className="container">
        <div className="row">
          <div className="col-md-8">
            {!profile ? (
              <Spinner />
            ) : (
              <div className="row">
                <div className="col-md-12 col-xs-12">
                  <section className="profile mt-4">
                    <div className="row">
                      <div className="col-md-3 offset-sm-1 col-sm-4 offset-md-3 col-2">
                        <img src={profile.image ? `${profile.image}` : "https://image.flaticon.com/icons/png/512/17/17004.png"} alt="" className="ourdp rounded img-fluid max-width: 100%; height: auto; " />
                      </div>
                      <div className="col-md-6 col-sm-7 col-10">
                        <h4>
                          <strong>{profile.username}</strong>
                          {status.isOwner ? (<Link to="/editprofile" className="btn ouredit btn-outline-dark">Edit Profile</Link>
                          ) : (
                            <ProfileButton
                              isProcessing={status.isProcessing} 
                              isFollowing={isFollowing}
                              unFollowUser={unFollowUser}
                              username={profile.username}
                            />
                          )
                          }
                        </h4>
                        <ProfileStats following={following} followers={followers} />
                        <p className="lead my-4">{profile.bio ? `${profile.bio}` : "No bio set"}</p>
                        <p>Bookmarked articles</p>
                        <ul className="list-group">
                          {
                            bookmarks.length > 0
                              ? bookmarks.map(bookmark => (
                                <li key={bookmark.article.slug} className="list-group-item"><a href={`article/${bookmark.article.slug}`}>{bookmark.article.title}</a></li>
                              ))
                              : <li className="list-group-item">You have no bookmarked articles</li>}
                        </ul>
                      </div>
                    </div>
                  </section>
                </div>
                <div className="col-md-2" />
              </div>
            )}
          </div>
          <Sidebar />
        </div>
      </div>
    </React.Fragment>
  );
};
export default ProfileView;
