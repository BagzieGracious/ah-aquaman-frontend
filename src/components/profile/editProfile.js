import React, { Component } from "react";
import { connect } from "react-redux";
import { updateProfile, loadProfile } from "../../actions/profile/profileActions";
import { storage } from "../../firebase";
import EditProfileView from "../../views/profile/editprofile";
import Navbar from "../article/navbarComponent/Navbar";
import ImageBanner from "../article/ImageBanner";


export class EditProfile extends Component {
  state = {
    image: null,
    url: "",
    progress: 0,
    bio: "",
    isUploading: null,
  };

  componentDidMount() {
    this.props.loadProfile();
    const { profile } = this.props.profile;
    const url = (profile && profile.image ? profile.image : "");
    this.setState({ url });
  }

  handleChange = (e) => {
    if (e.target.files[0]) {
      const image = e.target.files[0];
      this.setState(() => ({ image }));
      return this.handleUpload(image);
    }
  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  handleUpload = (image) => {
    const uploadTask = storage.ref(`images/${image.name}`)
      .put(image);
    uploadTask.on("state_changed",
      (snapshot) => {
        const isUploading = true;
        this.setState({ isUploading });
        const progress = Math.round((snapshot.bytesTransferred / snapshot.totalBytes) * 100);
        this.setState({ progress });
      },
      (error) => {
      },
      () => {
        // complete function ....
        /* istanbul ignore next */
        storage.ref("images")
          .child(image.name)
          .getDownloadURL()
          .then((url) => {
            this.props.updateProfile(url, "photo", this.props.history);
            this.setState({ url });
            const isUploading = false;
            this.setState({ isUploading });
          });
      });
  }

  onSubmit = (e) => {
    e.preventDefault();
    const { bio } = this.state;
    // Create user data
    const update = {
      bio,
    };
    this.props.updateProfile(update, "bio", this.props.history);
  };

  render() {
    const { isUpdating, profile, isLoading } = this.props.profile;
    return (
      <React.Fragment>
        <Navbar />
        <ImageBanner info="Edit Profile" page="profile" />
        <EditProfileView
          state={this.state}
          onChange={this.onChange}
          isUpdating={isUpdating}
          handleChange={this.handleChange}
          onSubmit={this.onSubmit}
          isLoading={isLoading}
          profile={profile}
        />
      </React.Fragment>
    );
  }
}

export const mapStateToProps = state => ({
  profile: state.profile,
});

export default connect(mapStateToProps, { updateProfile, loadProfile })(EditProfile);
