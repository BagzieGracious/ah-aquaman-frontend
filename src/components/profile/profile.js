import React from "react";
import { connect } from "react-redux";
import { loadProfile } from "../../actions/profile/profileActions";
import bookmarkList from "../../actions/bookmarksListAction";
import ProfileView from "../../views/profile/profile";
import Navbar from "../article/navbarComponent/Navbar";
import ImageBanner from "../article/ImageBanner";
import Sidebar from "../article/sidebar";
import {
  getUserfollowing, getUserfollowers, followUser, unFollowUser,
} from "../../actions/followUnfollow/followUnfollow";


export class Profile extends React.Component {
  componentDidMount() {
    const username = this.props.match &&  this.props.match.params.username?this.props.match.params.username:localStorage.getItem("username");
      this.props.loadProfile(username);
      this.props.getUserfollowing(username);
      this.props.getUserfollowers(username); 
      this.props.bookmarkList();
  }

  isOwner=() => (this.props.match && this.props.match.params.username ? localStorage.getItem("username") === this.props.match.params.username : true)

  unFollowUser=(user, should) => {
    { should ? this.props.unFollowUser(user) : this.props.followUser(user); }
  }

isFollowing=() => {
  let followed = false;
  if (this.props.followers.followers) {
    this.props.followers.followers.forEach((element) => {
      if (element.username === localStorage.getItem("username")) {
        followed = true;
      }
    });
  }
  return followed;
}

render() {
  const { isLoading, profile } = this.props.profile;

  const status = {
    isOwner: this.isOwner(),
    isProcessing: this.props.isProcessing,
  };
  return (
    <ProfileView
      status={status}
      unFollowUser={this.unFollowUser} 
      isFollowing={this.isFollowing()}
      following={this.props.following}
      followers={this.props.followers}
      profile={profile}
      bookmarks={this.props.bookmarks}
    />
  );
}
}

export const mapStateToProps = state => ({
  profile: state.profile,
  following: state.followUnfollow.following,
  bookmarks: state.bookmarkListReducer.bookmarks,
  followers: state.followUnfollow.followers,
  isProcessing: state.followUnfollow.isFollowingSomeone,
});

export default connect(mapStateToProps, {
  loadProfile, getUserfollowing, getUserfollowers, bookmarkList, followUser, unFollowUser,
})(Profile);
