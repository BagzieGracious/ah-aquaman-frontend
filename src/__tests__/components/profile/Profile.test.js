import React from "react";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import store from "../../../store/index";
import { loadProfile } from "../../../actions/profile/profileActions";
import Profile from "../../../components/profile/profile";

describe("profile component tests", () => {
  it("should not regress", () => {
    const wrapper = shallow(<Profile />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should dispatch load profile", () => {
    store.dispatch(loadProfile());
    expect(store.getState().profile).toEqual(
      {
        isLoading: true, isUpdating: null, msg: null, profile: null,
      },
    );
  });

  it("should render when mounted", () => {
    const wrapper = mount(
      <Provider store={store}>
        <BrowserRouter className="container">
          <Profile />
        </BrowserRouter>
      </Provider>,
    );
    const data = {
      isLoading: true, isUpdating: null, msg: null, profile: null,
    };
    wrapper.setProps({ profile: data });
  });
});
