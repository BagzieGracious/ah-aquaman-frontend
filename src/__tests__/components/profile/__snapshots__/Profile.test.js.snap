// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile component tests should not regress 1`] = `ShallowWrapper {}`;
